File: packages/protocol-parser/src/hexToSchema.ts
import { StaticAbiType, DynamicAbiType, schemaAbiTypes, staticAbiTypeToByteLength } from "@latticexyz/schema-type";
import { Hex, hexToNumber, sliceHex } from "viem";
import { Schema } from "./common";
import { InvalidHexLengthForSchemaError, SchemaStaticLengthMismatchError } from "./errors";

export function hexToSchema(data: Hex): Schema {
  if (data.length !== 66) {
    throw new InvalidHexLengthForSchemaError(data);
  }

  const staticDataLength = hexToNumber(sliceHex(data, 0, 2));
  const numStaticFields = hexToNumber(sliceHex(data, 2, 3));
  const numDynamicFields = hexToNumber(sliceHex(data, 3, 4));
  const staticFields: StaticAbiType[] = [];
  const dynamicFields: DynamicAbiType[] = [];

  for (let i = 4; i < 4 + numStaticFields; i++) {
    const schemaTypeIndex = hexToNumber(sliceHex(data, i, i + 1));
    staticFields.push(schemaAbiTypes[schemaTypeIndex] as StaticAbiType);
  }
  for (let i = 4 + numStaticFields; i < 4 + numStaticFields + numDynamicFields; i++) {
    const schemaTypeIndex = hexToNumber(sliceHex(data, i, i + 1));
    dynamicFields.push(schemaAbiTypes[schemaTypeIndex] as DynamicAbiType);
  }

  // validate static data length
  const actualStaticDataLength = staticFields.reduce((acc, fieldType) => acc + staticAbiTypeToByteLength[fieldType], 0);
  if (actualStaticDataLength !== staticDataLength) {
    console.warn(
      `Schema "${data}" static data length (${staticDataLength}) did not match the summed length of all static fields (${actualStaticDataLength}). ` +
        `Is \`staticAbiTypeToByteLength\` up to date with Solidity schema types?`
    );
    throw new SchemaStaticLengthMismatchError(data, staticDataLength, actualStaticDataLength);
  }

  return { staticFields, dynamicFields };
}

File: packages/services/pkg/protocol-parser/hexToSchema.go
package protocolparser

import (
	schematype "latticexyz/mud/packages/services/pkg/schema-type"
)

func HexToSchema(data string) Schema {
	if len(data) != 66 {
		panic(ErrInvalidHexLengthForSchema)
	}

	staticDataLength := HexToNumber(HexSlice(data, 0, 2))
	numStaticFields := HexToNumber(HexSlice(data, 2, 3))
	numDynamicFields := HexToNumber(HexSlice(data, 3, 4))

	staticFields := []schematype.SchemaType{}
	dynamicFields := []schematype.SchemaType{}

	for i := 4; i < 4+numStaticFields; i++ {
		staticFields = append(staticFields, schematype.SchemaType(HexToNumber(HexSlice(data, i, i+1))))
	}
	for i := 4 + numStaticFields; i < 4+numStaticFields+numDynamicFields; i++ {
		dynamicFields = append(dynamicFields, schematype.SchemaType(HexToNumber(HexSlice(data, i, i+1))))
	}

	// Validate static data length
	actualStaticData := 0
	for _, field := range staticFields {
		actualStaticData += schematype.StaticAbiTypeToByteLength(field)
	}
	if actualStaticData != staticDataLength {
		panic(ErrSchemaStaticLengthMismatch)
	}

	return Schema{
		StaticFields:  staticFields,
		DynamicFields: dynamicFields,
	}
}
